kubespawner/proxy.py

```diff
@@ -128,32 +149,26 @@ def __init__(self, *args, **kwargs):
             parent=self, namespace=self.namespace, labels=labels
         )
         self.endpoint_reflector = EndpointsReflector(
-            parent=self, namespace=self.namespace, labels=labels
+            self, namespace=self.namespace, labels=labels
+        )
+        await asyncio.gather(
+            self.ingress_reflector.start(),
+            self.service_reflector.start(),
+            self.endpoint_reflector.start(),
         )
 
-    def _set_k8s_client_configuration(self):
-        # The actual (singleton) Kubernetes client will be created
-        # in clients.py shared_client but the configuration
-        # for token / ca_cert / k8s api host is set globally
-        # in kubernetes.py syntax.  It is being set here
-        # and this method called prior to shared_client
-        # for readability / coupling with traitlets values
-        try:
-            kubernetes.config.load_incluster_config()
-        except kubernetes.config.ConfigException:
-            kubernetes.config.load_kube_config()
-        if self.k8s_api_ssl_ca_cert:
-            global_conf = client.Configuration.get_default_copy()
-            global_conf.ssl_ca_cert = self.k8s_api_ssl_ca_cert
-            client.Configuration.set_default(global_conf)
-        if self.k8s_api_host:
-            global_conf = client.Configuration.get_default_copy()
-            global_conf.host = self.k8s_api_host
-            client.Configuration.set_default(global_conf)
```

Comment:
I'm a bit confused about this section in general, and I'm looking to learn more in case it opens up a possibility to refactor this smarter.

1. What is the difference between `load_incluster_config` and `load_kube_config`?
2. Do we need to have the config loading happen before we do the latter part, or can we switch the ordering?

Reply:
`load_incluster_config` is what JupyterHub will use.  It relies on the pod's serviceaccount credentials, automatically mounted.  `load_kube_config` is what is used if you're using a .kubeconfig file, and is how running the test via pytest works.

I think we have to set the TLS stuff *after* loading the configuration, because I believe that loading the config overwrites all configuration settings, but I have not verified that.

Reply:
This isn't encouraging from upstream (granted, the sync client, but even so...):

https://github.com/kubernetes-client/python/issues/1005

like, everyone agrees it'd be nice to have a method that does the fallback, but no one has actually put it into the base client yet.

Reply:
I think I'm going to try the mixin class approach and move _await_async_init into there.